library/core/class.cookieidentity.php

```diff
@@ -98,18 +106,41 @@ public function getIdentity() {
             return $this->UserID;
         }
 
-        if (!$this->_checkCookie($this->CookieName)) {
-            $this->_clearIdentity();
-            return 0;
+        $userID = 0;
+        $name = $this->CookieName;
+        $version = $this->getCookieVersion($name);
+
+        switch ($version) {
+            case 1:
+                if ($this->_checkCookie($name)) {
+                    list($userID) = self::getCookiePayload($name);
+                    // Old cookie identity. Upgrade it.
+                    $this->setIdentity($userID);
+                } else {
+                    $this->_clearIdentity();
+                }
+                break;
+            case self::VERSION:
+            default:
+                $payload = $this->getJWTPayload($name);
+                if ($payload !== null) {
+                    Gdn::pluginManager()
+                        ->fireAs(self::class)
+                        ->fireEvent('getIdentity', ['payload' => &$payload]);
```

Comment:
Shouldn't that even be fire even when updating the the cookie too?

You could make setIdentity return the payload, loose the break and check is payload is not set before calling `$this->getJWTPayload($name);`.